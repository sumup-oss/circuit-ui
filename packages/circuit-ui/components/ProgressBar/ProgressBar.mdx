import { Meta, Status, Props, Story } from '../../../../.storybook/components';
import * as Stories from './ProgressBar.stories';

<Meta of={Stories} />

# ProgressBar

<Status.Stable />

A progress bar is a UI element used to visualise the progression of a task or operation. The visual element is always combined with a label describing a status value, either in percentage or fraction format.

<Story of={Stories.Steps} />
<Props />

## Usage guidelines

The progress bar should be used to indicate the amount of steps or time left to complete a task. It should not, however, be used to indicate the time needed to load a page. For a loading state, use the [Spinner component](Components/Spinner).

## Component variations

### Step or time-based

A progress bar can be used to display the progression of steps in a task or a fixed amount of time. The component adapts automatically based on the props it is passed: when you pass the `max` or `value` props, it renders statically. When you pass the `duration`, `pause`, or `loop` props, it smoothly animates the progress bar from empty to complete.

<Story of={Stories.Steps} />
<Story of={Stories.Timer} />

### Percentage or fraction label

When you specify the label for the progress bar, it should only be shown as either a fraction (e.g. _1/3_) or percentage (e.g. _33%_).

<Story of={Stories.Labelled} />

### Variants

You can use the progress bar in two variants: `primary` or `secondary`.

<Story of={Stories.Variants} />

### Sizes

You can use the progress bar in one of three sizes: `byte`, `kilo`, or `mega`.

<Story of={Stories.Sizes} />

## Accessibility

> This section is a stub. Contributions welcome!

- The ProgressBar component accepts a `hideLabel` prop that can be used to visually hide the required `label`. Use it with caution: if there is any doubt about what the progress bar could represent, it should have a visible label.
