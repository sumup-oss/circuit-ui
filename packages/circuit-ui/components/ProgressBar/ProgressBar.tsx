/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactNode } from 'react';
import { css, keyframes } from '@emotion/core';

import styled, { NoTheme, StyleProps } from '../../styles/styled';
import { hideVisually, textKilo } from '../../styles/style-mixins';
import { uniqueId } from '../../util/id';
import { ReturnType } from '../../types/return-type';

// TODO: Make the label prop required.
interface BaseProps {
  /**
   * @deprecated Use the `label` prop instead.
   */
  children?: ReactNode;
  /**
   * Choose from 2 style variants. Default: 'primary'.
   */
  variant?: 'primary' | 'secondary';
  /**
   * Choose from 3 sizes. Default: 'kilo'.
   */
  size?: 'byte' | 'kilo' | 'mega';
  /**
   * A descriptive label that is used by screenreaders.
   */
  label?: string;
  /**
   * Visually hide the label. It will remain accessible to screenreaders.
   */
  hideLabel?: boolean;
}

interface StepProgressProps {
  /**
   * A number greater than zero, indicating how much work the task requires.
   */
  max?: number;
  /**
   * A number between 0 and max, indicating how much of the task has been
   * completed.
   */
  value?: number;
}

interface TimeProgressProps {
  /**
   * The time it takes the progress bar to fill up in milliseconds.
   */
  duration?: number;
  /**
   * Whether the progress animation should loop indefinitely.
   */
  loop?: boolean;
  /**
   * Whether the animation should the paused.
   */
  paused?: boolean;
}

export type ProgressBarProps = BaseProps &
  StepProgressProps &
  TimeProgressProps;

type TaskProgressElProps = Omit<BaseProps, 'label'> & StepProgressProps;
type TimeProgressElProps = Omit<BaseProps, 'label'> & TimeProgressProps;
type LabelElProps = Pick<BaseProps, 'hideLabel'>;

const wrapperStyles = () => css`
  display: flex;
  align-items: center;
`;

const ProgressBarWrapper = styled('div')<NoTheme>(wrapperStyles);

const baseStyles = ({ theme }: StyleProps) => css`
  label: progress-bar;
  background-color: ${theme.colors.n200};
  border-radius: ${theme.borderRadius.pill};
  position: relative;
  width: 100%;
  overflow: hidden;

  &::after {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    background-color: ${theme.colors.p500};
    transition: width 0.05s ease-out;
    height: 100%;
    width: 1px;
  }
`;

const taskStyles = ({ value = 0, max = 1 }: StepProgressProps) => {
  const width = value && max ? (value / max) * 100 : 0;
  return css`
    &::after {
      width: ${width}%;
    }
  `;
};

const oneAnimation = keyframes`
  0% {
    width: 0%;
  }

  100% {
    width: 100%;
  }
`;

const loopAnimation = keyframes`
  0% {
    width: 0%;
    transform: translateX(0%);
  }

  90% {
    width: 100%;
    transform: translateX(0%);
  }

  100% {
    width: 100%;
    transform: translateX(100%);
  }
`;

const timedStyles = ({
  theme,
  duration = 3000,
  loop = false,
  paused = false,
}: StyleProps & TimeProgressElProps) => css`
  &::after {
    border-radius: ${theme.borderRadius.pill};
    animation-name: ${loop ? loopAnimation : oneAnimation};
    animation-duration: ${duration}ms;
    animation-play-state: ${paused ? 'paused' : 'running'};
    animation-fill-mode: forwards;
    animation-iteration-count: ${loop ? 'infinite' : 1};
  }
`;

const variantStyles = ({
  theme,
  variant = 'primary',
}: StyleProps & Omit<BaseProps, 'label'>) => {
  const variantMap = {
    primary: theme.colors.p500,
    secondary: theme.colors.n900,
  };
  return css`
    label: ${`progress-bar--${variant}`};

    &::after {
      background-color: ${variantMap[variant]};
    }
  `;
};

const sizeStyles = ({
  theme,
  size = 'kilo',
}: StyleProps & Omit<BaseProps, 'label'>) => {
  const sizeMap = {
    byte: theme.spacings.bit,
    kilo: theme.spacings.byte,
    mega: theme.spacings.mega,
  };
  return css({
    label: `progress-bar--${size}`,
    height: sizeMap[size],
  });
};

const TaskProgress = styled('span')<TaskProgressElProps>(
  baseStyles,
  variantStyles,
  sizeStyles,
  taskStyles,
);

const TimeProgress = styled('span')<TimeProgressElProps>(
  baseStyles,
  variantStyles,
  sizeStyles,
  timedStyles,
);

const labelStyles = ({ theme }: StyleProps) => css`
  label: progress-bar__label;
  ${textKilo({ theme })};
  margin-left: ${theme.spacings.byte};
`;

const labelHiddenStyles = ({ hideLabel }: LabelElProps) =>
  hideLabel &&
  css`
    label: progress-bar__label--hidden;
    ${hideVisually()};
  `;

const ProgressBarLabel = styled('span')<LabelElProps>(
  labelStyles,
  labelHiddenStyles,
);

/**
 * The ProgressBar component communicates the progress of a task or timer
 * to the user.
 */
export function ProgressBar(props: BaseProps & StepProgressProps): ReturnType;
export function ProgressBar(props: BaseProps & TimeProgressProps): ReturnType;
export function ProgressBar({
  children,
  max,
  value,
  size,
  variant,
  duration,
  loop,
  paused,
  label,
  hideLabel,
  ...props
}: ProgressBarProps) {
  const ariaId = uniqueId('progress-bar_');
  const title = hideLabel ? label : undefined;
  return (
    <ProgressBarWrapper {...props}>
      {max || value ? (
        <TaskProgress
          role="progressbar"
          title={title}
          aria-valuenow={value}
          aria-valuemin={0}
          aria-valuemax={max}
          aria-labelledby={ariaId}
          max={max}
          value={value}
          size={size}
          variant={variant}
        />
      ) : (
        <TimeProgress
          title={title}
          aria-labelledby={ariaId}
          duration={duration}
          loop={loop}
          paused={paused}
          size={size}
          variant={variant}
        />
      )}
      <ProgressBarLabel id={ariaId} hideLabel={hideLabel}>
        {children || label}
      </ProgressBarLabel>
    </ProgressBarWrapper>
  );
}
