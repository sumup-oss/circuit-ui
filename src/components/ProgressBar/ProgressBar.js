/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import { childrenPropType } from '../../util/shared-prop-types';
import { textKilo, subtractUnit } from '../../styles/style-helpers';
import { uniqueId } from '../../util/id';
import { sizes } from '../../styles/constants';

const { KILO, MEGA, GIGA } = sizes;

const calculateSize = ({ theme, size }) => {
  const sizeMap = {
    [KILO]: theme.spacings.byte,
    [MEGA]: theme.spacings.mega,
    [GIGA]: theme.spacings.tera
  };
  return sizeMap[size];
};

const wrapperStyles = ({ theme }) => css`
  label: progress-bar;
  display: flex;
  align-items: center;
  margin-bottom: ${theme.spacings.mega};
`;

const progressStyles = ({ theme, size, value, max }) => {
  const outerBorderWidth = '1px';
  const outerBorderRadius = theme.borderRadius.mega;
  const innerBorderRadiusLeft = `${subtractUnit(
    outerBorderRadius,
    outerBorderWidth
  )}`;
  const innerBorderRadiusRight =
    value && max && (value / max) * 100 === 100 ? innerBorderRadiusLeft : 0;
  const width = value && max ? (value / max) * 100 : 0;
  return css`
    label: progress-bar__progress;
    background-color: ${theme.colors.n100};
    border: ${outerBorderWidth} solid ${theme.colors.n300};
    border-radius: ${outerBorderRadius};
    position: relative;
    width: 100%;
    height: ${calculateSize({ theme, size })};
    overflow: hidden;

    &::after {
      content: '';
      display: block;
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      width: ${width}%;
      background-color: ${theme.colors.p500};
      border: 1px solid ${theme.colors.p700};
      box-shadow: inset 0 1px 0 0 ${theme.colors.p300};
      border-radius: ${innerBorderRadiusLeft} ${innerBorderRadiusRight}
        ${innerBorderRadiusRight} ${innerBorderRadiusLeft};
      transition: width 0.05s ease-out;
    }
  `;
};

const labelStyles = ({ theme }) => css`
  label: progress-bar__label;
  ${textKilo({ theme })};
  margin-left: ${theme.spacings.byte};
`;

const ProgressBarWrapper = styled('div')`
  ${wrapperStyles};
`;
const ProgressBarProgress = styled('span')`
  ${progressStyles};
`;
const ProgressBarLabel = styled('span')`
  ${labelStyles};
`;

/**
 * Progress bar component to indicate progress
 */
const ProgressBar = ({ children, max, value, ...props }) => {
  const ariaId = uniqueId('progress-bar_');
  return (
    <ProgressBarWrapper>
      <ProgressBarProgress
        role="progressbar"
        aria-valuenow={value}
        aria-valuemin="0"
        aria-valuemax={max}
        aria-labelledby={ariaId}
        {...{ ...props, max, value }}
      />
      <ProgressBarLabel id={ariaId}>{children}</ProgressBarLabel>
    </ProgressBarWrapper>
  );
};

ProgressBar.KILO = KILO;
ProgressBar.MEGA = MEGA;
ProgressBar.GIGA = GIGA;

ProgressBar.propTypes = {
  /**
   * A number greater than zero, indicating how much work the task requires.
   */
  max: PropTypes.number,
  /**
   * A number between 0 and max, indicating how much of the task has been
   * completed.
   */
  value: PropTypes.number,
  /**
   * Size
   */
  size: PropTypes.oneOf([ProgressBar.KILO, ProgressBar.MEGA, ProgressBar.GIGA]),
  /**
   * Child nodes to be rendered as the label.
   */
  children: childrenPropType
};

ProgressBar.defaultProps = {
  size: ProgressBar.KILO,
  max: 1.0,
  value: 0,
  children: null
};

/**
 * @component
 */
export default ProgressBar;
