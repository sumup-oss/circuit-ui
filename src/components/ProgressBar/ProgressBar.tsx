/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement, ReactNode } from 'react';
import { css, keyframes } from '@emotion/core';

import styled, { StyleProps } from '../../styles/styled';
import { textKilo } from '../../styles/style-helpers';
import { uniqueId } from '../../util/id';

interface BaseProps {
  children?: ReactNode;
  /**
   * Choose from 2 style variants. Default: 'primary'.
   */
  variant?: 'primary' | 'secondary';
  /**
   * Choose from 3 sizes. Default: 'kilo'.
   */
  size?: 'byte' | 'kilo' | 'mega';
}

interface StepProgressProps {
  /**
   * A number greater than zero, indicating how much work the task requires.
   */
  max?: number;
  /**
   * A number between 0 and max, indicating how much of the task has been
   * completed.
   */
  value?: number;
}

interface TimeProgressProps {
  /**
   * The time it takes the progress bar to fill up in milliseconds.
   */
  duration?: number;
  /**
   * Whether the progress animation should loop indefinitely.
   */
  loop?: boolean;
  /**
   * Whether the animation should the paused.
   */
  paused?: boolean;
}

export type ProgressBarProps = BaseProps &
  StepProgressProps &
  TimeProgressProps;

type ReturnType = ReactElement<any, any> | null;

const wrapperStyles = () => css`
  display: flex;
  align-items: center;
`;

const ProgressBarWrapper = styled('div')<{}>(wrapperStyles);

const baseStyles = ({ theme }: StyleProps) => css`
  label: progress-bar;
  background-color: ${theme.colors.n200};
  border-radius: ${theme.borderRadius.pill};
  position: relative;
  width: 100%;
  overflow: hidden;

  &::after {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    background-color: ${theme.colors.p500};
    transition: width 0.05s ease-out;
    height: 100%;
    width: 1px;
  }
`;

const taskStyles = ({ value = 0, max = 1 }: StepProgressProps) => {
  const width = value && max ? (value / max) * 100 : 0;
  return css`
    &::after {
      width: ${width}%;
    }
  `;
};

const oneAnimation = keyframes`
  0% {
    width: 0%;
  }

  100% {
    width: 100%;
  }
`;

const loopAnimation = keyframes`
  0% {
    width: 0%;
    transform: translateX(0%);
  }

  90% {
    width: 100%;
    transform: translateX(0%);
  }

  100% {
    width: 100%;
    transform: translateX(100%);
  }
`;

const timedStyles = ({
  duration = 3000,
  loop = false,
  paused = false
}: TimeProgressProps) => css`
  &::after {
    animation-name: ${loop ? loopAnimation : oneAnimation};
    animation-duration: ${duration}ms;
    animation-play-state: ${paused ? 'paused' : 'running'};
    animation-fill-mode: forwards;
    animation-iteration-count: ${loop ? 'infinite' : 1};
  }
`;

const variantStyles = ({
  theme,
  variant = 'primary'
}: StyleProps & BaseProps) => {
  const variantMap = {
    primary: theme.colors.p500,
    secondary: theme.colors.n900
  };
  return css`
    label: ${`progress-bar--${variant}`};

    &::after {
      background-color: ${variantMap[variant]};
    }
  `;
};

const sizeStyles = ({ theme, size = 'kilo' }: StyleProps & BaseProps) => {
  const sizeMap = {
    byte: theme.spacings.bit,
    kilo: theme.spacings.byte,
    mega: theme.spacings.mega
  };
  return css({
    label: `progress-bar--${size}`,
    height: sizeMap[size]
  });
};

const TaskProgress = styled('span')<BaseProps & StepProgressProps>(
  baseStyles,
  variantStyles,
  sizeStyles,
  taskStyles
);

const TimeProgress = styled('span')<BaseProps & TimeProgressProps>(
  baseStyles,
  variantStyles,
  sizeStyles,
  timedStyles
);

const labelStyles = ({ theme }: StyleProps) => css`
  label: progress-bar__label;
  ${textKilo({ theme })};
  margin-left: ${theme.spacings.byte};
`;

const ProgressBarLabel = styled('span')<{}>(labelStyles);

/**
 * The ProgressBar component communicates the progress of a task or timer
 * to the user.
 */
export function ProgressBar(props: BaseProps & StepProgressProps): ReturnType;
export function ProgressBar(props: BaseProps & TimeProgressProps): ReturnType;
export function ProgressBar({
  children,
  max,
  value,
  size,
  variant,
  duration,
  loop,
  paused,
  ...props
}: ProgressBarProps) {
  const ariaId = uniqueId('progress-bar_');
  return (
    <ProgressBarWrapper {...props}>
      {max || value ? (
        <TaskProgress
          role="progressbar"
          aria-valuenow={value}
          aria-valuemin={0}
          aria-valuemax={max}
          aria-labelledby={ariaId}
          max={max}
          value={value}
          size={size}
          variant={variant}
        />
      ) : (
        <TimeProgress
          aria-labelledby={ariaId}
          duration={duration}
          loop={loop}
          paused={paused}
          size={size}
          variant={variant}
        />
      )}
      <ProgressBarLabel id={ariaId}>{children}</ProgressBarLabel>
    </ProgressBarWrapper>
  );
}
