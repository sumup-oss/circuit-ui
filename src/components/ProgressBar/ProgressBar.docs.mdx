import { Status, Props, Story } from '../../../.storybook/components';
import ProgressBar from './ProgressBar';

# Progress bar

<Status.Stable />

A progress bar is a UI element used to visualise the progression of a task or
operation. The visual element is always combined with a label describing a
status value either in percentage or fraction format.

<Story id="components-progressbar--base" />

<Props of={ProgressBar} />

## When to use it

The progress bar should be used to indicate the amount of time left to complete
a task. It should not however, be used to indicate the time needed to load a page.

## Usage guidelines

- **Do** place the status value to the right of the progress bar.
- **Do** use in moderation, only for tasks, operations or information that may take a long time.
- **Do not** use an indeterminate value beyond percentage or fraction.
- **Do not** use a progress bar for every task, operation or information.

## Component variations

### Percentage or fraction completion

When you specify the label for the progress bar, it should only be shown
as either a fraction (e.g. _1/3_) or percentage (e.g. _33%_).

<Story id="components-progressbar--with-fraction" />

<Story id="components-progressbar--with-percentage" />

### Sizes

You can use the progress bar in a variety of sizes: `ProgressBar.KILO`, `ProgressBar.MEGA`, or `ProgressBar.GIGA`.

<Story id="components-progressbar--size" />
