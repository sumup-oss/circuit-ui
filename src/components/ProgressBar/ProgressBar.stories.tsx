/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/** @jsx jsx */
import { Fragment } from 'react';
import { css, jsx } from '@emotion/core';

import docs from './ProgressBar.docs.mdx';
import { ProgressBar, ProgressBarProps } from './ProgressBar';

export default {
  title: 'Components/ProgressBar',
  component: ProgressBar,
  parameters: {
    docs: { page: docs },
  },
};

const variants = ['primary', 'secondary'] as const;
const sizes = ['byte', 'kilo', 'mega'] as const;

const progressBarStyles = css`
  width: 90%;
  min-width: 500px;
`;

export const Steps = (args: ProgressBarProps) => (
  <ProgressBar {...args} css={progressBarStyles} />
);

Steps.args = {
  value: 3,
  max: 10,
};

export const Timer = (args: ProgressBarProps) => (
  <ProgressBar {...args} css={progressBarStyles} />
);

Timer.args = {
  duration: 3000,
  paused: false,
  loop: true,
  max: null,
};

export const Labelled = (args: ProgressBarProps) => {
  const fraction = `${args.value}/${args.max}`;
  const percentage = `${(args.value! / args.max!) * 100}%`;
  return (
    <Fragment>
      <ProgressBar {...args} css={progressBarStyles}>
        {fraction}
      </ProgressBar>
      <ProgressBar {...args} css={progressBarStyles}>
        {percentage}
      </ProgressBar>
    </Fragment>
  );
};

Labelled.args = {
  value: 3,
  max: 10,
};

export const Variants = (args: ProgressBarProps) =>
  variants.map((variant) => (
    <ProgressBar
      key={variant}
      {...args}
      variant={variant}
      css={progressBarStyles}
    >
      {variant}
    </ProgressBar>
  ));

Variants.args = {
  value: 3,
  max: 10,
};

export const Sizes = (args: ProgressBarProps) =>
  sizes.map((size) => (
    <ProgressBar key={size} {...args} size={size} css={progressBarStyles}>
      {size}
    </ProgressBar>
  ));

Sizes.args = {
  value: 3,
  max: 10,
};
