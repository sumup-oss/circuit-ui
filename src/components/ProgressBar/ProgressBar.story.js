/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Fragment } from 'react';
import { select, number } from '@storybook/addon-knobs/react';
import { css } from '@emotion/core';

import docs from './ProgressBar.docs.mdx';
import ProgressBar from './ProgressBar';

export default {
  title: 'Components/ProgressBar',
  component: ProgressBar,
  parameters: {
    docs: { page: docs },
    jest: ['ProgressBar']
  }
};

const progressBarStyles = css`
  min-width: 500px;
  max-width: 90%;
`;

export const base = () => {
  const size = select(
    'Size',
    {
      kilo: ProgressBar.KILO,
      mega: ProgressBar.MEGA,
      giga: ProgressBar.GIGA
    },
    ProgressBar.KILO
  );
  const max = number('Maximum value', 10);
  const value = number('Value', 3);
  return (
    <ProgressBar value={value} max={max} size={size} css={progressBarStyles} />
  );
};

export const withFraction = () => {
  const max = 10;
  const value = 7;
  const children = `${value}/${max}`;
  return (
    <ProgressBar value={value} max={max} css={progressBarStyles}>
      {children}
    </ProgressBar>
  );
};

export const withPercentage = () => {
  const max = 10;
  const value = 7;
  const children = `${(value / max) * 100}%`;
  return (
    <ProgressBar value={value} max={max} css={progressBarStyles}>
      {children}
    </ProgressBar>
  );
};

export const size = () => {
  const max = 10;
  const value = 7;
  const children = `${(value / max) * 100}%`;
  return (
    <Fragment>
      <ProgressBar
        size={ProgressBar.KILO}
        value={value}
        max={max}
        css={progressBarStyles}
      >
        {children}
      </ProgressBar>
      <ProgressBar
        size={ProgressBar.MEGA}
        value={value}
        max={max}
        css={progressBarStyles}
      >
        {children}
      </ProgressBar>
      <ProgressBar
        size={ProgressBar.GIGA}
        value={value}
        max={max}
        css={progressBarStyles}
      >
        {children}
      </ProgressBar>
    </Fragment>
  );
};
