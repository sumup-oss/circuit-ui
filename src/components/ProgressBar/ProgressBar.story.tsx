/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { select, number, boolean } from '@storybook/addon-knobs';
import { css } from '@emotion/core';

import docs from './ProgressBar.docs.mdx';
import { ProgressBar, ProgressBarProps } from './ProgressBar';

export default {
  title: 'Components/ProgressBar',
  component: ProgressBar,
  parameters: {
    docs: { page: docs },
    jest: ['ProgressBar']
  }
};

const variants = ['primary', 'secondary'] as const;
const sizes = ['byte', 'kilo', 'mega'] as const;

const BaseProgressBar = (props: ProgressBarProps) => (
  <ProgressBar
    variant={select('Variant', variants, 'primary')}
    size={select('Size', sizes, 'kilo')}
    css={css`
      min-width: 500px;
      max-width: 90%;
    `}
    {...props}
  />
);

export const Steps = () => (
  <BaseProgressBar
    value={number('Value', 3)}
    max={number('Maximum value', 10)}
  />
);

export const Timer = () => (
  <BaseProgressBar
    duration={number('Duration', 3000)}
    paused={boolean('Paused', false)}
    loop={boolean('Loop', true)}
  />
);

export const Labelled = () => {
  const max = number('Maximum value', 10);
  const value = number('Value', 3);
  const fraction = `${value}/${max}`;
  const percentage = `${(value / max) * 100}%`;
  return (
    <>
      <BaseProgressBar value={value} max={max}>
        {fraction}
      </BaseProgressBar>
      <BaseProgressBar value={value} max={max}>
        {percentage}
      </BaseProgressBar>
    </>
  );
};

export const Variants = () => {
  const max = number('Maximum value', 10);
  const value = number('Value', 3);
  return variants.map(variant => (
    <BaseProgressBar key={variant} variant={variant} value={value} max={max}>
      {variant}
    </BaseProgressBar>
  ));
};

export const Sizes = () => {
  const max = number('Maximum value', 10);
  const value = number('Value', 3);
  return sizes.map(size => (
    <BaseProgressBar key={size} size={size} value={value} max={max}>
      {size}
    </BaseProgressBar>
  ));
};
