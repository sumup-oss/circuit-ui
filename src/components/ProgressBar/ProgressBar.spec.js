import React from 'react';

import ProgressBar from '.';

describe('ProgressBar', () => {
  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(<ProgressBar />);
    expect(actual).toMatchSnapshot();
  });

  it('should render with progress styles', () => {
    const actual = create(<ProgressBar value={0.5} />);
    expect(actual).toMatchSnapshot();
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(<ProgressBar />);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
