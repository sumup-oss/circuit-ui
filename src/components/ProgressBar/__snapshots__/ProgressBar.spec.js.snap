// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar should render with default styles 1`] = `
.circuit-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 16px;
}

.circuit-0 {
  background-color: #FAFBFC;
  border: 1px solid #D8DDE1;
  border-radius: 4px;
  position: relative;
  width: 100%;
  height: 8px;
  overflow: hidden;
}

.circuit-0::after {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  width: 0%;
  background-color: #3388FF;
  border: 1px solid #1760CE;
  box-shadow: inset 0 1px 0 0 #AFD0FE;
  border-radius: calc(4px - 1px) 0 0 calc(4px - 1px);
  -webkit-transition: width 0.05s ease-out;
  transition: width 0.05s ease-out;
}

.circuit-2 {
  font-size: 13px;
  line-height: 20px;
  margin-left: 8px;
}

<div
  class="circuit-4 circuit-5"
>
  <span
    aria-labelledby="progress-bar_1"
    aria-valuemax="1"
    aria-valuemin="0"
    aria-valuenow="0"
    class="circuit-0 circuit-1"
    max="1"
    role="progressbar"
    value="0"
  />
  <span
    class="circuit-2 circuit-3"
    id="progress-bar_1"
  />
</div>
`;

exports[`ProgressBar should render with progress styles 1`] = `
.circuit-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 16px;
}

.circuit-2 {
  font-size: 13px;
  line-height: 20px;
  margin-left: 8px;
}

.circuit-0 {
  background-color: #FAFBFC;
  border: 1px solid #D8DDE1;
  border-radius: 4px;
  position: relative;
  width: 100%;
  height: 8px;
  overflow: hidden;
}

.circuit-0::after {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  width: 50%;
  background-color: #3388FF;
  border: 1px solid #1760CE;
  box-shadow: inset 0 1px 0 0 #AFD0FE;
  border-radius: calc(4px - 1px) 0 0 calc(4px - 1px);
  -webkit-transition: width 0.05s ease-out;
  transition: width 0.05s ease-out;
}

<div
  class="circuit-4 circuit-5"
>
  <span
    aria-labelledby="progress-bar_2"
    aria-valuemax="1"
    aria-valuemin="0"
    aria-valuenow="0.5"
    class="circuit-0 circuit-1"
    max="1"
    role="progressbar"
    value="0.5"
  />
  <span
    class="circuit-2 circuit-3"
    id="progress-bar_2"
  />
</div>
`;
